draft-bbs-signatures.md

```diff
@@ -366,37 +375,44 @@ Procedure:
 
 1. (W, H0, H) = octets_to_point(PK)
 
-2. h = XOF(SK  || msg[i] || ... || msg[L])
+2. sig_dst = HASH(PK || L || H_s || H_d || H_1 || ... || H_L || CipherInfo)
```

Comment:
Need to specify the form of `PK` and other parameters e.g octets_to_point

Reply:
Apologies, ignore saw the definition above r.e PK is the existing definition for L etc?

Reply:
That's a good point. L is implicitly defined as the total number of signed messages (based on that the input to sign is (msg_1, ..., msg_L)). I did notice though that the definition of L in terminology is different. Updated the definition of L to "the total number of signed messages" (there are still some inconsistencies with the definition of L even beyond this PR. I will submit a different PR to fix them).

Reply:
Ok but do we need to convert L to bytes here? using I2OSP? Same for the other parameters too we need to make sure we fully define the form they are in to compute this hash

Reply:
That's true. Note that the same consideration holds for the challenge hash in spkGen/verify. Also, the current approach is consistent with what the [Schnorr NIZK proofs RFC](https://datatracker.ietf.org/doc/html/rfc8235#section-2.3) does. That said, i would like to be more specific but i don't really like adding `point_to_octets` for every argument (for aesthetic reasons only). Another option could be a note that specifies the correct transformations for each element to be hashed (`point_to_octets` for points, `I2OSP` for numbers). I added such note in the definition to the HASH function in the Parameters section for review.

Reply:
I like your current approach which is to make the conversion implicit in the definition of the HASH function.

Reply:
I also noted this recommendation in the RFC you referenced is this something we should consider?

```
Within the hash function, there must be a clear boundary between any
   two concatenated items.  It is RECOMMENDED that one should always
   prepend each item with a 4-byte integer that represents the byte
   length of that item.  OtherInfo may contain multiple subitems.  In
   that case, the same rule shall apply to ensure a clear boundary
   between adjacent subitems.
```

Reply:
Thank you! Yes I think we should. I opened up an [issue](https://github.com/decentralized-identity/bbs-signature/issues/96) to discuss some of the options. I will make another PR to move the spec to this direction.

Reply:
See PR #97 for a proposal moving towards this direction. The addition to the definition of HASH is removed from this PR and re-introduced extended to #97.

Reply:
Thank you! I added some text to reflect that both H_s and H_d are mandatory, in their definitions in Terminology. Does this address your review?

Reply:
In regards to adding the length to hashed items, I do think it is necessary in a couple cases but not for all elements.

The current formula is `sig_dst = HASH(PK || L || H_s || H_d || H_1 || ... || H_L || CipherInfo)`. Here the encoded length of the PK and the generator points is known for the associated curve, and `L` serves to disambiguate the number of following generator points (perhaps it should be L+2 here). Prepending CipherInfo with its length helps to avoid any length extension attacks (although I don't believe SHAKE-256 suffers from that issue). If we added ApplicationInfo after CipherInfo, then adding a length to each would also help avoid any confusion around the boundaries between the two.

Reply:
Thank you! Agreed. IMO it should be scalars (like the messages), integers (like L) and generic octet strings (like cipherInfo) that need to be prepended with their length (see PR #97 and issue #96 also). Regarding L specifically i do agree it may not be strictly necessary in `spk_dst` since it comes after PK and is followed by H_s, both of which have well defined lengths, but i think in general non-negative integers should be prepended with their length as a good practice.

(note that this is not only to avoid length extension which even if it was possible i don't think it is that dangerous against `spk_dst`, since the hash value would most likely change under this attack)

I also think i prefer `L` to `L+2` for aesthetic reason mainly and to more clearly reflect the number of signed messages. Since `H_s` and `H_d` are always supplied, essentially what might change is `L` either way.

Reply:
Scalars would also be encoded with a fixed length, and integers can be by defining the width of the field in the spec (f.ex. 4 bytes is plenty for the number of signed messages). For things like the revealed messages values in the challenge, that can be prepended by the number of following elements, each element being a pair of a 4-byte integer position and scalar value. In any case this is probably over-thinking things, but I suppose it doesn't hurt to be careful.

Reply:
We're getting into encoding layers and I don't think we should. That should be avoided. I2OSP and OSP2I are fine since they specify the length in bytes but that's all we should be using.

Reply:
@mikelodder7 we have to sufficiently describe the encoding of elements as they pertain to operations within the scheme to promote interoperability

Reply:
I also feel that describing the formats of the messages to be hashed would be advantageous and is consistent with what other specs that deal with this stuff do. But I do also agree that going into too much detail and be too much restrictive is not a good thing. The minimum specification to achieve security must be preferred.

Reply:
> Scalars would also be encoded with a fixed length, and integers can be by defining the width of the field in the spec (f.ex. 4 bytes is plenty for the number of signed messages). For things like the revealed messages values in the challenge, that can be prepended by the number of following elements, each element being a pair of a 4-byte integer position and scalar value. In any case this is probably over-thinking things, but I suppose it doesn't hurt to be careful.

When it comes to integers I agree and i updated PR #96 accordingly. When it comes to scalars there is two options i see,

1. Their length will be defined by the ciphersuite. This means that confusion regarding the used ciphersuite may lead to confusion regarding the scalars length.
2. The spec will define a fixed xLen to be used by all ciphersuites as an input to I2OSP.

IMO 2. seems too restrictive. Also spec extensions may define a different value for xLen which would lead to the same problem as option 1. For that reason, i prefer 1. and the proposed approach to prepend the lengths to scalars, although i don't feel that strongly about it and i would also go with option 2. or another proposal.

Reply:
If there is confusion over the ciphersuite being used then there would be big problems, that must be validated either through the DST or the generator creation process. I think we just say that scalars have a fixed length encoding defined by the ciphersuite.

Reply:
That's true. Well personally I would prefer even when ciphersuites define the length of the scalar, to be more careful and also enforce it here even if it is redundant, but im fine with the alternative. I will also be very interested in @christianpaquin input on that issue, since UProve seems to have taken this route.

Reply:
> I will also be very interested in @christianpaquin input on that issue, since UProve seems to have taken this route.

Just one thing I'd like to point out: currently the `H_d` generator is multiplied with a random `s` value. In U-Prove, we set that message value to the hash of various elements (issuer params, and an always-disclosed "token information" attribute). This allows issuers to encode a message that should always be disclosed (e.g., validity period, signature metadata, usage restrictions, etc.) when presenting the signature, without "wasting" another message for that purpose if needed. Maybe something to consider here?

Reply:
Thank you very much for your input! That is interesting. In bbs+ the `s` value is never revealed (the paper goes to great lengths actually to keep it secret, see issue #70  for some context). This is to avoid correlation. And from the conversation in the issue i linked to, it seems that there is no getting around the fact that we need `s` to be random and we need it to be secret. That said, it is a very interesting idea and definitely worth thinking more about it.